apis/placement/v1beta1/stageupdate_types.go

```diff
@@ -281,15 +281,23 @@ type StageConfig struct {
 	// +kubebuilder:validation:Optional
 	// +kubebuilder:validation:XValidation:rule="!self.exists(e, e.type == 'Approval' && has(e.waitTime))",message="AfterStageTaskType is Approval, waitTime is not allowed"
 	// +kubebuilder:validation:XValidation:rule="!self.exists(e, e.type == 'TimedWait' && !has(e.waitTime))",message="AfterStageTaskType is TimedWait, waitTime is required"
-	AfterStageTasks []AfterStageTask `json:"afterStageTasks,omitempty"`
+	AfterStageTasks []StageTask `json:"afterStageTasks,omitempty"`
+
+	// The collection of tasks that needs to completed successfully by each stage before starting the stage.
+	// Each task is executed in parallel and there cannot be more than one task of the same type.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:MaxItems=1
+	// +kubebuilder:validation:XValidation:rule="!self.exists(e, e.type == 'Approval' && has(e.waitTime))",message="AfterStageTaskType is Approval, waitTime is not allowed"
+	// +kubebuilder:validation:XValidation:rule="!self.exists(e, e.type == 'TimedWait')",message="BeforeStageTaskType cannot be TimedWait"
+	BeforeStageTasks []StageTask `json:"beforeStageTasks,omitempty"`
 }
 
-// AfterStageTask is the collection of post-stage tasks that ALL need to be completed before moving to the next stage.
-type AfterStageTask struct {
-	// The type of the after-stage task.
+// StageTask is the pre or post stage task that needs to be completed before starting or moving to the next stage.
+type StageTask struct {
+	// The type of the before or after stage task.
 	// +kubebuilder:validation:Enum=TimedWait;Approval
 	// +kubebuilder:validation:Required
```

Comment:
How do you handle the fact that you are adding a new property that is required?
I would expect this to fail get+re-put of existing updateruns?
at least any client that does not bump the SDK will fail manipulating the updatrun object, as it won't be able to serialize the Type property?
Is there a way to configure defaults on added properties? is there a way to test this?

Reply:
I think that this will work as existing updateruns would have no value, so it will get a default value on read of "TimedWait", which is the correct expectation.
new updateruns are required to specify the type because of validation.